serve/publish.go

```diff
@@ -0,0 +1,180 @@
+	return os.WriteFile(outputPath, buffer.Bytes(), 0644)
+}
+
+func (p *PluginServe) build(ctx context.Context, pluginDirectory string, goos string, goarch string) error {
+	pluginName := "plugin" + "_" + goos + "_" + goarch
+	distPath := pluginDirectory + "/dist"
+
+	pluginPath := distPath + "/" + pluginName
+	args := []string{"build", "-C", pluginDirectory, "-o", pluginPath}
+	cmd := exec.Command("go", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = os.Environ()
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to build plugin with `go %v`: %w", args, err)
+	}
+
+	pluginFile, err := os.Open(pluginPath)
+	if err != nil {
+		return fmt.Errorf("failed to open plugin file: %w", err)
+	}
+	defer pluginFile.Close()
+
+	zipPluginPath := pluginPath + ".zip"
+	zipPluginFile, err := os.Create(zipPluginPath)
+	if err != nil {
+		return fmt.Errorf("failed to create zip file: %w", err)
+	}
+	defer zipPluginFile.Close()
+
+	zipWriter := zip.NewWriter(zipPluginFile)
+	defer zipWriter.Close()
+
+	pluginZip, err := zipWriter.Create(pluginName)
+	if err != nil {
+		return fmt.Errorf("failed to create file in zip archive: %w", err)
+	}
+	if err := os.Remove(pluginPath); err != nil {
+		return fmt.Errorf("failed to remove plugin file: %w", err)
+	}
+	_, err = io.Copy(pluginZip, pluginFile)
+	if err != nil {
+		return fmt.Errorf("failed to copy plugin file to zip archive: %w", err)
+	}
+	return nil
+}
+
+func (s *PluginServe) writeManifest(ctx context.Context, dir string) error {
+	manifest := Manifest{
+		Name:             s.plugin.Name(),
+		Version:          s.plugin.Version(),
+		Title:            s.plugin.Title(),
+		ShortDescription: s.plugin.ShortDescription(),
+		Description:      s.plugin.Description(),
+		Categories:       s.plugin.Categories(),
+		Protocols:        s.versions,
+		SupportedTargets: s.plugin.Targets(),
+		PackageType:      PackageTypeNative,
+	}
+	buffer := &bytes.Buffer{}
+	m := json.NewEncoder(buffer)
+	m.SetIndent("", "  ")
+	m.SetEscapeHTML(false)
+	err := m.Encode(manifest)
+	if err != nil {
+		return err
+	}
+	outputPath := filepath.Join(dir, "plugin.json")
+	return os.WriteFile(outputPath, buffer.Bytes(), 0644)
+}
+
+func (s *PluginServe) newCmdPluginPublish() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "publish <plugin_directory>",
+		Short: pluginPublishShort,
+		Long:  pluginPublishLong,
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			pluginDirectory := args[0]
+			distPath := pluginDirectory + "/dist"
+			if isDirectoryExist(distPath) {
+				return fmt.Errorf("dist directory already exist: %s", distPath)
+			}
+			if err := os.MkdirAll(distPath, 0755); err != nil {
+				return err
+			}
+			if err := s.plugin.Init(cmd.Context(), nil, plugin.NewClientOptions{
+				NoConnection: true,
+			}); err != nil {
+				return err
+			}
+			if err := s.writeTablesJson(cmd.Context(), distPath); err != nil {
```

Comment:
There's a bit of overlap here with the CLI tables command. Not sure what to do about it yet, just stating it.
Also this is for our internal usage so maybe better to have it separated.

I would create a `schema.json` for both the manifest and tables so we can validate those. We'll need to be aware that changing/adding things in the Column/Table structs can break the JSON